### DISCUSSION TITLE
Phase out functionality slowly in new versions

### DISCUSSION THREAD

**Discussion by darius-m** [keycloak/keycloak#14492] [Q&A]
Version 19.0.2 completely disables **all** Javascript mappers for SAML clients. The [release notes](https://www.keycloak.org/2022/09/keycloak-1902-released.html#_deploying_scripts_through_saml_javascript_protocol_mapper) mention that going forward the ability to upload scripts through the admin console is disabled; however, the wording does not imply that existing mappers would also be removed. Since this is a breaking change, there should actually be emphasis added to it.

The script upload ability should be slowly phased out - the [server developer guide](https://www.keycloak.org/docs/latest/server_development/#_script_providers) actually mentions the ability to re-enable the functionality using the `upload_scripts` feature that does not appear to work with Quarkus (written either as `upload-scripts` or `upload_scripts`).

@mposolda, @stianst, please advise how the old scripts can be recovered.

**Comment by darius-m** [comment]
On an additional note, the `scripts` feature that the new functionality appears to rely on is shown as `preview`, which feels like a regression.

**Reply by darius-m** [reply]
For anyone interested, I think I got to the expected functionality by doing the following:
  - get a list of Javascript mappers and their scripts by querying the database - get the scripts from the `PROTOCOL_MAPPER_CONFIG` table (match the `PROTOCOL_MAPPER_ID` field to the ID of mappers in the `PROTOCOL_MAPPER` table that have `javascript` in their respective `PROTOCOL_MAPPER_NAME` field);
  - enable the `scripts` feature in Keycloak;
  - inside an empty directory create the `META-INF` directory and a set of Javascript files (one for each mapper):
  - `META-INF/keycloak-scripts.json` contains a list of SAML mappers (besides other optional files, as mentioned in the server developer guide) - for example:
```json
{
  "saml-mappers": [
    {
      "name": "Javascript - Prefixed Username",
      "fileName": "prefixed-username.js",
      "description": "Exports the username with a prefix"
    }
  ]
}
```
  - the Javascript files contain the Javascript code for the mappers (one per mapper; the same scripts can be used as before) - for example:
```javascript
"u-" + user.getUsername();
```

All files can then be added inside a jar (zip) file and moved to the `providers/` directory inside Keycloak - **Note:** Keycloak will need to be built and restarted afterwards:
```bash
$  zip -r saml-mappers.jar META-INF *.js
$  cp saml-mappers.jar ${KEYCLOAK_DIR}/providers/
$  ${KEYCLOAK_DIR}/bin/kc.sh build
$  systemctl restart keycloak
```

Afterwards, when Keycloak loads the mappers, they will appear as mapper categories when using the `Create` option in the admin interface.

**Note**: There are still some issues with this approach - the old mappers are still in the database, but they do not show in the admin interface so they cannot be easily deleted; because of this, the new mappers will need to have a different name. I am not sure how to delete the old mappers cleanly, or if they can cause some issues with duplicated exported data in some instances.

**Reply by stianst** [reply]
> For anyone interested, I think I got to the expected functionality by doing the following:
>   - get a list of Javascript mappers and their scripts by querying the database - get the scripts from the `PROTOCOL_MAPPER_CONFIG` table (match the `PROTOCOL_MAPPER_ID` field to the ID of mappers in the `PROTOCOL_MAPPER` table that have `javascript` in their respective `PROTOCOL_MAPPER_NAME` field);
>   - enable the `scripts` feature in Keycloak;
>   - inside an empty directory create the `META-INF` directory and a set of Javascript files (one for each mapper):
>   - `META-INF/keycloak-scripts.json` contains a list of SAML mappers (besides other optional files, as mentioned in the server developer guide) - for example:
> ```json
> {
>   "saml-mappers": [
>     {
>       "name": "Javascript - Prefixed Username",
>       "fileName": "prefixed-username.js",
>       "description": "Exports the username with a prefix"
>     }
>   ]
> }
> ```
>   - the Javascript files contain the Javascript code for the mappers (one per mapper; the same scripts can be used as before) - for example:
> ```javascript
> "u-" + user.getUsername();
> ```
> 
> All files can then be added inside a jar (zip) file and moved to the `providers/` directory inside Keycloak - **Note:** Keycloak will need to be built and restarted afterwards:
> ```bash
> $  zip -r saml-mappers.jar META-INF *.js
> $  cp saml-mappers.jar ${KEYCLOAK_DIR}/providers/
> $  ${KEYCLOAK_DIR}/bin/kc.sh build
> $  systemctl restart keycloak
> ```
> 
> Afterwards, when Keycloak loads the mappers, they will appear as mapper categories when using the `Create` option in the admin interface.
> 
> **Note**: There are still some issues with this approach - the old mappers are still in the database, but they do not show in the admin interface so they cannot be easily deleted; because of this, the new mappers will need to have a different name. I am not sure how to delete the old mappers cleanly, or if they can cause some issues with duplicated exported data in some instances.

Thanks for this. We should probably turn this into a section in the migration docs. We should also find some way of cleaning up the existing script mappers in the db.

**Comment by stianst** [comment]
Hi, we deprecated the ability to upload scripts through the admin endpoints quite a while back, which was then finally removed in 18 or 19.

Somehow though during this deprecation the SAML protocol mappers where forgotten and they where not disabled like the other parts. The fact that continued working was a defect, rather than expected behaviour.

Due to the fact that we consider uploading scripts through the admin endpoints a security risk and that the SAML protocol mappers where left in by mistake we decided these would have to be removed quickly.

I appreciate we may not have the best communication around this though and will try to improve on that in the future.

**Reply by darius-m** [reply]
Thank you for the answer. I understand the reasoning, but I was mostly confused by the wording. The changes mention that the admins "should" upload scripts through the server directly. The word "should" implies that there is an option to not use the approach (the word "must" has to be used when a strong requirement is in place).

As such, I assumed that any scripts that had been uploaded to the server through the admin console would still be available (and they are in fact still in the database), but read-only, and the option to add new scripts would be disabled. This approach would offer a longer transition period, and would have the added benefit of offering the possibility to delete the old mappers when the scripts are moved to the jar file. I would assume that this would have to be removed at some point, since it increases the amount of code, but Keycloak appears to have a reasonable approach to phasing out in most cases.

As a bit of feedback on the user side of things, I think that the mappers should be moved to the `Add predefined mapper` section instead, if the scripts do not have any configurable parameters. It would also make sense to be able to see the mapper's code (read-only) to figure out their functionality if the user believes this is needed. I can open a feature request, but I am not sure if it should be open on this repository or https://github.com/keycloak/keycloak-ui.

**Comment by Foobartender** [comment]
Is it safe to delete leftover mappers manually from the database tables protocol_mapper and protocol_mapper_config?
